docs/language-reference/04-types.md

```diff
@@ -0,0 +1,323 @@
+> This convention means that wherever Slang refers to "rows" or "columns" of a matrix, the equivalent terms in the GLSL, SPIR-V, OpenGL, and Vulkan specifications are "column" and "row" respectively (*including* in the compound terms of "row-major" and "column-major")
+> While it may seem that this choice of convention is confusing, it is necessary to ensure that subscripting with `[]` can be efficiently implemented on all target platforms.
+> This decision in the Slang language is consistent with the compilation of HLSL to SPIR-V performed by other compilers.
+
+### Legacy Syntax
+
+For compatibility with older codebases, the generic `matrix` type includes default values for `T`, `R`, and `C`, being declared as:
+
+```hlsl
+struct matrix<T = float, let R : int = 4, let C : int = 4> { ... }
+```
+
+This means that the bare name `matrix` may be used as a type equivalent to `float4x4`:
+
+```hlsl
+// All of these variables have the same type
+matrix a;
+float4x4 b;
+matrix<float, 4, 4> c;
+```
+
+Structure Types
+---------------
+
+Structure types are introduced with `struct` declarations, and consist of an ordered sequence of named and typed fields:
+
+```hlsl
+struct S
+{
+    float2 f;
+    int3 i;
+}
+```
+
+### Standard Layout
+
+The _standard layout_ for a structure type uses the following algorithm:
+
+* Initialize variables `size` and `alignment` to zero and one, respectively
+* For each field `f` of the structure type:
+  * Update `alignment` to be the maximum of `alignment` and the alignment of `f`
+  * Set `size` to the smallest multiple of `alignment` not less than `size`
+  * Set the offset of field `f` to `size`
+  * Add the size of `f` to `size`
+
+When this algorithm completes, `size` and `alignment` will be the size and alignment of the structure type.
+
+Most target platforms do not use the standard layout directly, but it provides a baseline for defining other layout algorithms.
+Any layout for structure types must guarantee an alignment at least as large as the standard layout.
+
+### C-Style Layout
+
+C-style layout for structure types differs from standard layout by adding an additional final step:
+
+* Set `size` the smallest multiple of `alignment` not less than `size`
+
+This mirrors the layout rules used by typical C/C++ compilers.
+
+### D3D Constant Buffer Layout
+
+D3D constant buffer layout is similar to standard layout with two differences:
+
+* The initial alignment is 16 instead of one
+
+* If a field would have _improper straddle_, where the interval `(fieldOffset, fieldOffset+fieldSize)` (eclusive on both sides) contains any multiple of 16, *and* the field offset is not already a multiple of 16, then the offset of the field is adjusted to the next multiple of 16
+
+Array Types
+-----------
+
+An _array type_ is either a statically-sized or dynamically-sized array type.
```

Comment:
Good point about arrays being value types. There's a related point there that because arrays are value types there is no specific reason why they can't/shouldn't be used as function return types, but I don't recall if we actually added the IR pass required to make that Just Work.

As for passing an array of unknown size to a function, its the kind of thing that should work, but I don't think we'd implemented all the required IR infrastructure. Ideally a function should be able to define a parameter that is an array of unknown size, and then you can call that function with either known-size or unknown-size arrays.

At the IR level, the simplistic approach would be to treat each function parameter of unknown array size as introducing an implicit generic `int` parameter that gives the size of the array, and then specializing functions based on actual array sizes at call sites. That approach could of course lead to heavy code bloat.

A smarter approach is probably to assume that at the "mechanism" level unknown-size arrays get passed by `const` reference, but with a convention that the callee needs to make their own copy of the array if they want to do mutation. Actually implementing that copy becomes a problem for any target that doesn't have an `alloca()` equivalent, so it isn't really a full solution...

Maybe the simplest answer is to decide by fiat that function parameters that have an unknown-size array type are immutable (at which point they can be passed by-value or by-const-reference and there should be no difference).